### PR TITLE
feat: initialize database

### PR DESCRIPTION
**Pull request by ArnabXD**
Example usage

```ts
import db from '../../db'

async function testknex (){
        if (!(await knex.schema.hasTable('pmpermit'))) {
		await knex.schema.createTable('pmpermit', table => {
			table.increments('id')
			table.string('user_id', 25)
			table.boolean('approved').defaultTo(false)
		})
	}
}

testknex()
```

Requires a new Env variable `DATABASE_URI` with postgres connection string as value

### CODE REVIEW

**Review by roj1512 — COMMENTED** [review]
We don’t want the user to be forced to use a database. This should be backward-compatible.

**Comment by ArnabXD** [comment]
Now the program won't quit if the database is not set. Do you recommend any other way that I didn't understand previously?

File: src/knex.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { knex as Knex } from 'knex'
+import env from './env'
+
+export default Knex({
+	client: 'pg',
+	connection: env.DATABASE_URI
+})
```
Comment (roj1512) [inline src/knex.ts:7]: Won’t this error if `connection` is `''`?

Comment (ArnabXD) [reply src/knex.ts:7]: I just tested it and it doesn't cause problem as long as I don't import it anywhere, need to fix...

File: src/knex.ts [hunk 2]
```diff
@@ -1,7 +1,23 @@
...
-	connection: env.DATABASE_URI
-})
+class SingletonKnex {
+	private static instance: Knex
+	private constructor() {
+		//
+	}
+	public static getInstance(): Knex {
+		if (!env.DATABASE_URI) {
+			return <Knex>{}
```
Comment (ArnabXD) [inline src/knex.ts:11]: If I throw an Error here it crashes the program on launch. What is your opinion on this @roj1512 ?

Comment (roj1512) [reply src/knex.ts:11]: I don’t know. Why do you call this method on launch?

**Review by ArnabXD — COMMENTED** [review]
Usage :

```js
import Knex from '../../knex'

...
// inside plugin handler functions
const knex = Knex.getInstance()
```

**Comment by roj1512** [comment]
I am unsure if we need a class.

File: src/knex.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+		}
+		if (!SingletonKnex.instance) {
+			SingletonKnex.instance = _knex({
+				client: 'pg',
+				connection: env.DATABASE_URI
+			})
+		}
+		return SingletonKnex.instance
+	}
+}
```
Comment (roj1512) [inline src/knex.ts:21]:
```suggestion
let instance: Knex | undefined

export default function knex() {
	if (!env.DATABASE_URI) {
		hrow new Error('DATABASE_URI variable not found')
	}
	if (instance == undefined) {
		instance =  _knex({
			client: 'pg',
			connection: env.DATABASE_URI
		})
	}
	return instance
}
```

Comment (ArnabXD) [reply src/knex.ts:21]: Yes, this can be done both with class and function. whichever you prefer.

**Comment by roj1512** [comment]
Hmm

File: src/knex.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+		throw new Error('DATABASE_URI variable not found')
+	}
+	if (instance === undefined) {
+		instance = _knex({
+			client: 'pg',
+			connection: env.DATABASE_URI
+		})
+	}
+	return instance
+}
```
Comment (roj1512) [inline src/knex.ts:16]:
```suggestion
let instance: Knex | undefined

const {knex} = new class {
	get knex() {
		if (!env.DATABASE_URI) {
		throw new Error('DATABASE_URI variable not found')
	}
	if (instance === undefined) {
		instance = _knex({
			client: 'pg',
			connection: env.DATABASE_URI
		})
	}
	return instance
	}
}

export default knex
```

**Review by roj1512 — COMMENTED** [review]
Maybe I was wrong about classes. What if we use a class this way? Will it work? If it does, we will no longer be required to use parathesis. `knex()` would be `knex`.

File: src/knex.ts (hunk 4 above)
Comment (ArnabXD) [reply src/knex.ts:16]: this approach doesn't crash for importing, but using it in a CommandHandler function crashes the app on launch.

Comment (roj1512) [reply src/knex.ts:16]: Can I see an example handler that crashes?

**Review by roj1512 — APPROVED** [review]

**Pull request merged**